libnest/src/transaction/remove.rs

```diff
@@ -53,63 +47,79 @@ impl RemoveTransaction {
 
         let npf_explorer = NPFExplorer::from(&npf_path).map_err(|_| InvalidCachedPackageFile)?;
 
-        let instructions_handle = npf_explorer
-            .load_instructions()
-            .map_err(|_| InvalidCachedPackageFile)?;
+        remove_package(config, lock_ownership, npf_explorer, self.target())
+    }
+}
 
-        if let Some(executor) = &instructions_handle {
-            executor
-                .execute_before_remove(config.paths().root())
-                .map_err(PreRemoveInstructionsFailure)?;
-        }
+fn is_empty_directory(dir_path: &Path) -> std::io::Result<bool> {
+    let mut it = fs::read_dir(dir_path)?;
 
-        // If the package is effective, installed files must be removed
-        if npf_explorer.manifest().kind() == Kind::Effective {
-            // Open the log file, and remove all the files listed in it
-            let log = config
-                .installed_packages_cache(lock_ownership)
-                .package_log(self.target())
-                .map_err(LogFileLoadError)?;
-
-            // Iterate backwards to ensure removal of nested files before that of top-level directories
-            for entry in log.files().into_iter().rev() {
-                let abs_path = Path::new("/").with_content(entry.path());
-                let rel_path = config.paths().root().with_content(entry.path());
-
-                if let Ok(metadata) = fs::symlink_metadata(&rel_path) {
-                    match (entry.file_type().is_dir(), metadata.file_type().is_dir()) {
-                        // The file to remove is a directory, remove it if it is empty
-                        (true, true) => {
-                            if let Ok(true) = is_empty_directory(&rel_path) {
-                                fs::remove_dir(&rel_path)
-                            } else {
-                                Ok(())
-                            }
-                        }
+    Ok(it.next().is_none())
+}
 
-                        // The file was expected to be a directory, but is a symlink, leave it
-                        (true, false) if metadata.file_type().is_symlink() => Ok(()),
+/// Remove the package from a given [`NPFExplorer`], using a given [`PackageID`]'s log
+pub(crate) fn remove_package(
```

Comment:
`extract_package` has its own module because it is used by both `Upgrade` and `Install` transactions, whereas `remove_package` is only used by the `Remove` transaction.
I'll leave it as it is for now.

Reply:
No, it is also used by the `Upgrade` transaction, which emphasis the imbalance described above :p

Reply:
Merf, that's right. Still, the `remove_package` function should obviously belong to the `remove` module, otherwise it would be even weirder.

Maybe the right choice would be to rename the `RemoveTransaction` to `UninstallTransaction` (it feels more appropriate) and move it into its own module, but it is way beyond the scope of this PR, which is why I said earlier I'd leave it as is (and do the renaming later).